Updated python highlights

doubleloop opened the pull request:
Probably best to review each commit separately.

vigoux commented:
Thanks for contributing !
LGTM, let's have know the views of @theHamsta on this too, as he is the language maintainer.

vigoux commented on lua/nvim-treesitter/highlight.lua:
```diff
@@ -48,6 +49,7 @@ hlmap["type"] = "Type"
 hlmap["type.builtin"] = "Type"
 hlmap["structure"] = "Structure"
 hlmap["include"] = "Include"
+hlmap["variable"] = "Identifier"
```
i don't really agree with this, as "variable" will be the default highlighting for everything then.
What is your opinion on this @kyazdani42 ?

kyazdani42 replied:
well `Identifier` is defined as `any variable name` in vim documentation. not sure what you mean by 'the default highlighting

theHamsta replied:
Now, everything is red with my colorscheme. "Everything" should by rather neutral for my taste, so I would need to make `variable` to no hihglight after this change. How does this @vigoux @kyazdani42 look on your computers? If this change would make the highlighting unusable for most colorschemes we should not do it.

kyazdani42 replied:
basically identifier should not be used for variables as red is definitively not a good color. Identifier applies well on vimscript, but not on most languages, specially not python

vigoux commented on lua/nvim-treesitter/highlight.lua:
```diff
@@ -13,6 +13,7 @@ local hlmap = vim.treesitter.TSHighlighter.hl_map
 hlmap.error = "Error"
 hlmap["punctuation.delimiter"] = "Delimiter"
 hlmap["punctuation.bracket"] = "Delimiter"
+hlmap["punctuation.special"] = "Delimiter"
```
You might want to add this to the `CONTRIBUTING.md` file too.

theHamsta replied:
Also add a highlight `punctuation.bracket` for normal `{` and `}` (e.g. for dicts)

theHamsta commented on queries/python/highlights.scm:
```diff
@@ -3,31 +3,33 @@
-
 ((identifier) @constant
- (match? @constant "^[A-Z][A-Z_]*$"))
+ (match? @constant "^[A-Z][A-Z_0-9]*$"))
```
Could you add a built-in highlight for `assert` and a highlight for `__dunder__` identifiers?

theHamsta commented:
You should also squash your commits.

doubleloop commented:
>The only issue that I have is that variable highlight. I would prefer to remove the (identifier) @variable query.

the issue I wanted to solve wit 538a697 was this highlight:
![Screenshot from 2020-06-03 00-06-55](https://user-images.githubusercontent.com/2602750/83575519-f6bf4e80-a52f-11ea-80de-53e5a2a72928.png)

(here I use orange for `Identifier`, green for `Function` and yellow for `String`)

![Screenshot from 2020-06-03 00-06-31](https://user-images.githubusercontent.com/2602750/83575525-f921a880-a52f-11ea-8c7f-d5b51a1ed1d4.png)

Issues here:
1. `foo` in line 1 should have the same color as foo inside format string
1. It would be desirable to be able to customize color of `foo` in line 1, and because `:h group-name` states that::
>         *Identifier         any variable name
if not `Identifier`, what is reasonable choice here?

kyazdani42 commented:
i guess normal would be more appropriate. Most colorscheme decide to make identifier red because it matches a lot of stuff in vim (and tags for instance)

theHamsta commented:
We need to make sure that the unspecific `(identifier) @variable` does not override more specific highlights. How is the precedence of highlighting handled right now. Right now @variable was overwriting the built-in highlight of `assert`. This you also be a bug of neovim or by the order we specify the highlights.

vigoux commented:
I see now, and that's a tricky one.
I think that we can find a way to fix that using those `@embedded` captures you see in some of atom queries.
I don't really know how we could do this but as we investigate language injection (which is the next item in my todo list after https://github.com/neovim/neovim/pull/12321 and updating the queries), we will dive a bit into how we can use multiple parsers, and highlight things correctly.

theHamsta commented:
Hi. I guess with @variable on top of the file everything is working as it should (`assert` is not overwritten). No special treatment of @embedded is necessary.

The only open question is how @variable should be hightlighted. I am open for any decision but that decision should be propageted to the other languages. 

Thanks again for improving the highlights!

theHamsta commented:
Could you check why constanst no longer get highlighted properly with your PR?

Without PR
![image](https://user-images.githubusercontent.com/7189118/83743578-32a30280-a65b-11ea-9b5a-ce51c347fa86.png)

With PR:
![image](https://user-images.githubusercontent.com/7189118/83743620-42224b80-a65b-11ea-9769-71486969a03b.png)

theHamsta commented on queries/python/highlights.scm:
```diff
@@ -169,41 +170,23 @@
-((attribute
-    object: (identifier) @type)
-    (match? @type "^[A-Z][a-z_]+"))
-
```
Why delete those?

vigoux commented:
>  No special treatment of `@embedded` is necessary.

If I remenber how python works, `f"4 = {var}"` are interpolated strings, and thus a whole new tree can appear within `{}`, which requires `@embedded`

theHamsta commented:
In theory yes, but in practice the stuff contained interpolated string is already contained in the main-AST.

At least, it works out-of-the-box right now. But I didn't try out really complicated expressions.

vigoux commented:
A better way to do it then : 
```
(string) @string
;; Reset highlighting for string interpolation
(interpolation) @Normal

;; Other highlighting afterward
```

theHamsta commented:
`@embedded` will rock for C macros! I already put a TODO in the C highlights.

(those poor Github users with `@embedded` and `@variable` handles! They get mentioned all the time you forget to escape the `@something`)

theHamsta commented:
I'm getting the correct hightlights with your PR if I move the following queries further down the file.

- use  "^[A-Z][A-Z_\\d]+$" for constants like in the other repos

```
((identifier) @type
 (match? @type "^[A-Z]"))
((identifier) @constant
 (match? @constant "^[A-Z][A-Z_\\d]+$"))
```

![image](https://user-images.githubusercontent.com/7189118/84326080-76769a00-ab7c-11ea-9897-fb4a6b61a379.png)

After that and squashing this PR should be ready to go. Except @kyazdani42 @vigoux "@variable" should `Normal`? I am ok with `Identifier`. It's as fine as `Normal` for me.

vigoux commented:
Is	`@varlable` still needed if you do the modification I suggested above?
Having `@variable` to Normal is the same as ignoring it.

theHamsta commented:
> Is `@varlable` still needed if you do the modification I suggested above?
> Having `@variable` to Normal is the same as ignoring it.
Except that it fixes the F-string issue.

vigoux commented:
Quick side note: for now the `#match?` predicate uses vim regexes (in very-magic mode), so you can use vim's character classes if needed

theHamsta commented:
@doubleloop do you want to finish this? or should we just wrap up your changes?

doubleloop commented:
@theHamsta sry for late response, I was afk

>Could you check why constanst no longer get highlighted properly with your PR

I fixed it in #a4f1081 . I used negation (although vim regex has strange syntax for this), to avoid repetitions (for example uppercase identifiers are constants so instead of writing this rule everywhere we make more specific rule, this rules can have exceptions for constants)

> You should also squash your commits.

If the fixups are ok, I will squash all.

> We need to make sure that the unspecific (identifier) @variable does not overwrite more specific highlights. How is the precedence of highlighting handled right now?

AFAIK if there are conflicting for the highlight, the last one is used. Thus if more specific cases are defined below, it will be ok.

It would be nice if ordering would not matter and this could also work:
```
(call
  function: (identifier) @function)
(identifier) @Normal
```
but I guess it would require some changes in vim.

> With this PR `@variable` was overwriting the built-in highlight of assert

It is not unless on top

> I don't really agree with this, as "variable" will be the default highlighting for everything

I am guessing that interpolation did not work because in case of no rule, vim probably used te color of previous token instead of resetting it. If this is true, setting default highlighting of identifier to Normal at the top of scm file is sane workaround IMO.

Still, naming this default as `variable` was bad idea (misleading and too specific) so I linked it directly to `@Normal` and removed it from `hlmap`

theHamsta commented:
Well, I still don't know how but I had the fealing that the queries coming first are painted before the ones coming after that. For me, it made a huge difference in which order I ordered the queries. Not sure, if this is desired or just an effect of implementation details in Neovim.

The official treesitter queries have both `(identifier) @variable` and rules for more specific identifiers. Not sure, whether they prioritize by the order of the queries or if they have a priority of the `@` highlights.

vigoux commented:
Highlighting priority is on its way within neovim, and the priority we are coming for is : the latest match within the query wins.
We can call that painting order if you want : latest applied wins.

doubleloop commented:
Just squashed the commits

theHamsta commented:
@vigoux apparently I have no write-access to merge it.

vigoux commented:
> @vigoux apparently I have no write-access to merge it.

It should be okay now

theHamsta commented:
Yeah!

vigoux commented:
Welcome to the team!!
Can we open an issue regarding the debate we had here though?
Some kind of rfc to flesh out how we are going to tackle `@embedded` and `@injection`?

stsewd commented:
Thanks everyone! I've been waiting for this!
